### DISCUSSION TITLE
tutorials/io/background_loading

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#286] [User-contributed notes]
# tutorials/io/background_loading

Commonly, games need to load resources asynchronously. When switching the main scene of your game (e.g. going to a new level), you might want to show a loading screen with some indication that prog...

https://docs.godotengine.org/en/stable/tutorials/io/background_loading.html

**Comment by terraquad** [comment]
The example doesn't demonstrate the use of `ResourceLoader.load_threaded_get_status()`. It should probably be checked in the `_on_button_pressed()` function before calling `ResourceLoader.load_threaded_get()`

**Reply by TranquilMarmot** [reply]
A more complete example getting the progress percentage with a progress bar:

```gdscript
extends Control
class_name LoadingScreen

## This is emitted when the scene is finished loading.
## Use `ResourceLoader.load_threaded_get(path)` to get the scene.
signal scene_loaded(path: String)

@onready var progress_bar = %ProgressBar

## The path to the scene that's actually being loaded
var path: String

## Actual progress value; we move towards towards this
var progress_value := 0.0


## Load the scene at the given path.
## When this is finished loading, the "scene_loaded" signal will be emitted.
func load_scene(path_to_load: String):
	path = path_to_load
	ResourceLoader.load_threaded_request(path)


func _process(delta: float):
	if not path:
		return

	var progress = []
	var status = ResourceLoader.load_threaded_get_status(path, progress)

	if status == ResourceLoader.ThreadLoadStatus.THREAD_LOAD_IN_PROGRESS:
		progress_value = progress[0] * 100
		progress_bar.value = move_toward(progress_bar.value, progress_value, delta * 20)

	if status == ResourceLoader.ThreadLoadStatus.THREAD_LOAD_LOADED:
		# zip the progress bar to 100% so we don't get weird visuals
		progress_bar.value = move_toward(progress_bar.value, 100.0, delta * 150)

		# "done" loading :)
		if progress_bar.value >= 99:
			scene_loaded.emit(path)
```

**Reply by mintymelinda** [reply]
I had to use `$ProgressBar` instead of `%ProgressBar` but this works beautifully thanks.

**Reply by sweetbbak** [reply]
you have to mark the node as "Access as Unique Name" to use %NodeName which will allow you to reference a node without knowing its exact path. Its pretty handy so that you don't have to waste time fixing hard coded paths if you move something around. You could always reference child nodes to a unique node pretty easily

**Reply by Viktorjb** [reply]
Not sure if this will help anyone, but after struggling to write a similar code in C# I realised that the enums THREAD_LOAD_IN_PROGRESS and THREAD_LOAD_LOADED are simply called InProgress and Loaded in C#.
Took me a while to figure that out and it doesn't seem to be mentioned anywhere in these docs.

**Reply by workpire-vii** [reply]
The code in the first reply does not work because load() is a built-in method and the engine won't work right with it. I just renamed it to loader() and it works fine. 4.4

**Reply by TranquilMarmot** [reply]
@workpire-vii thanks for calling that out - I just edited it to be `load_scene` instead of `load`

**Comment by Linshiqi** [comment]
why load_threaded_get_status use a array as progress which is a one-element array containing the ratio of completion of the threaded loading (between 0.0 and 1.0), shouldn't it better a variable?

**Reply by sweetbbak** [reply]
AFAIK you can load multiple Resources asynchronously at once. So the return type being an array allows you to get the loading percentage for any number of these resources that you are loading.

**Reply by Reneator** [reply]
also this enables to return more information via modifying the referenced resource.

Alternatively a "data-holder" object could be created/returned that could be asked for the loading-progress for the specific path.

But maybe there are some reasons why its better the way it currently is.

**Comment by Luminoth** [comment]
If instead of string path, this was an exported PackedScene, does that negate the ability to background load resources? In other words if I have something like this:

```
[Export]
private PackedScene _myLevel;

[...]

ResourceLoader.LoadThreadedRequest(_myLevel.ResourcePath);
```

is that actually loading the resource (non-async) earlier when the owning node is created?

Something else I've been wondering around this - where does the cost of Instantiate() come in? Like if I use this to async load a PackedScene and then I have to Instantiate / AddChild that scene, that's a whole other potential cost that you might want hidden behind a loading screen?

**Reply by Reneator** [reply]
@Luminoth 

Im in a similar situation i think, where the loading of the packagedScenes and the resources is only like 40% of loading time required, and a lot of stuff is happening on _ready() (initializing the UI Elements etc.) where the game can stay frozen for like 5-10 more seconds until the game is actually shown. From my research it seems that its not possible to run those functionalities asynchronously. (please anyone correct me if im wrong, i would love to be able to do this) So the screen will freeze until those _ready functionalities are completed.

But its already a much better situation, where the player now has to stare at the loading screen a bit longer, instead of having to stare at the engine loading-texture for like 30 seconds.

**Comment by tecanec** [comment]
So, what happens if you call `load_threaded_request` without a corresponding `load_threaded_get`? Does the resource just stay in memory until the next time you try to load it instead of getting freed by refcounting?

**Reply by XandaPanda42** [reply]
If you don't use load_threaded_get() on the resource, the ResourceLoader singleton will just hold on to the object forever as far as I can tell. When ResourceLoader finishes loading an object, it increases the reference count on the object, so I think it'd just stay in memory indefinitely. 

I don't think it'd be an issue (memory-wise) except in extreme circumstances, like forgetting to call load_threaded_get() on thousands of resources at once. Still, probably something to keep an eye out for.
